src/internal/group_api/mod.rs

```diff
@@ -351,6 +384,63 @@ pub(crate) fn get_group_keys<'a>(
     )
 }
 
+fn compare_users<T: Eq + std::hash::Hash + std::fmt::Debug, X>(
+    desired_users: &Vec<T>,
+    found_users: HashMap<T, X>,
+) -> IronOxideErr {
+    let desired_users_set: HashSet<&T> = HashSet::from_iter(desired_users);
+    let found_users_vec: Vec<T> = found_users.into_iter().map(|(x, _)| x).collect();
+    let found_users_set: HashSet<&T> = HashSet::from_iter(&found_users_vec);
+    let diff: Vec<&&T> = desired_users_set.difference(&found_users_set).collect();
+    IronOxideErr::UserDoesNotExist(format!("Failed to find the following users: {:?}", diff))
+}
+
+fn collect_admin_and_member_info<CR: rand::CryptoRng + rand::RngCore>(
+    recrypt: &Recrypt<Sha256, Ed25519, RandomBytes<CR>>,
+    signing_key: &crate::internal::DeviceSigningKeyPair,
+    group_priv_key: recrypt::api::PrivateKey,
+    admins: Vec<UserId>,
+    members: Vec<UserId>,
+    user_ids_and_keys: HashMap<UserId, PublicKey>,
+) -> Result<
+    (
+        Vec<(UserId, PublicKey, TransformKey)>,
```

Comment:
I think generating the transform keys in this function is not good separation of concerns. You could argue that if the number of members is huge there could be a performance reason to mix this in here so that you don't have to go over the list again, but I don't see that as a legitimate concern.

Reply:
I see this function as getting everything ready to make `request::GroupMember`s and `request::GroupAdmin`s, so generating the `TransformKey`s here seems to fulfill that purpose to me. Are you thinking of it more as just a way to re-partition the admins and members?

Reply:
Maybe we should rename the function to note that, or add a comment to it? Really it wouldn't be _that_ bad to have this return the 2 vecs and feed the members vec back into another function which would generate the transform keys. At the same time, this does break it up more than it was, which is incremental improvement.

I'll let you and clint figure out what needs to be done. :running_man:

Reply:
I still do think it would be cleaner to do the transform key generation as a separate step, but this solution isn't bad. I think the function signature makes it clear that transform keys are also being computed in here, but maybe putting a small doc comment on this function is a good idea anyway.